custom_components/miele/sensor.py

```diff
@@ -540,10 +610,62 @@ def device_state_attributes(self):
         return attributes
 
 
-class MieleTimeSensor(MieleRawSensor):
-    def __init(self, hass, device, key):
-        pass
+class MieleConsumptionSensor(MieleSensorEntity):
+    def __init__(self, hass, device, key, measurement):
+        super().__init__(hass, device, key)
+
+        self._attr_native_unit_of_measurement = measurement
+        self._cached_consumption = -1
+        self._attr_state_class = STATE_CLASS_TOTAL_INCREASING
+
+        if key == "energyConsumption":
+            self._attr_device_class = DEVICE_CLASS_ENERGY
 
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        device_state = self._device["state"]
+        device_status_value = self._device["state"]["status"]["value_raw"]
+
+        if not _is_running(device_status_value):
+            self._cached_consumption = -1
+            return 0
+
+        if self._cached_consumption >= 0:
+            if (
+                "ecoFeedback" not in device_state
+                or device_state["ecoFeedback"] is None
+                or device_status_value == STATUS_NOT_CONNECTED
+            ):
+                return self._cached_consumption
```

Comment:
@kloknibor I'm getting spurious resets and double energy consumption with the current release again. I'm wondering if the refactoring done could have broken it a little bit.

What was the reason for switching the order of this `if` block and `if not _is_running(device_status_value)`? In my PR, this is what it looked like:

```python

        if self._cached_consumption >= 0:
            if ("ecoFeedback" not in device_state or
                device_state["ecoFeedback"] is None or
                "currentEnergyConsumption" not in device_state["ecoFeedback"] or
                device_status_value == STATUS_NOT_CONNECTED):
                # Sometimes the Miele API seems to return a null ecoFeedback
                # object even though the Miele device is running. Or if the the
                # Miele device has lost the connection to the Miele cloud, the
                # status is "not connected". Either way, we need to return the
                # last known value until the API starts returning something
                # sane again, otherwise the statistics generated from this
                # sensor would be messed up.
                return self._cached_consumption

        if not _is_running(device_status_value):
            self._cached_consumption = -1
            return 0
```

After this PR, the check for the cached value is done after the `_is_running` check.

I know for a fact that each run of the dishwasher uses 0.7 kWh, and here's a screenshot from today from the energy dashboard, when the dishwasher has actually only been run once:
<img width="299" alt="bild" src="https://user-images.githubusercontent.com/1417619/141515267-796a43a0-9217-4bed-83ae-092e8a0de847.png">

And in the history graph of the consumption, the sudden reset can clearly be seen:

<img width="404" alt="bild" src="https://user-images.githubusercontent.com/1417619/141515375-b910c20f-36a0-4524-b47e-fcd8fef3de54.png">

Reply:
I flipped the order of the `if` statements locally now. I'll report back in a few days if I'm still seeing the spurious resets.

Reply:
The dishwasher has been running twice since, the first time it unfortunately looks like a spurious jump was reported :unamused: So it doesn't seem like my suggested change solves it, at least not completely.

I'll see if I can find anything else in the coming days.

![image](https://user-images.githubusercontent.com/1417619/142248833-e5ade302-dcb2-4aad-9821-d90beba0f94a.png)

Reply:
Hmm, just me or has the Miele API stopped reporting energy consumption? I haven't got any energy consumption in my HA sensors the past 3 days at least.

Reply:
> Hmm, just me or has the Miele API stopped reporting energy consumption? I haven't got any energy consumption in my HA sensors the past 3 days at least.

Sorry for late report as I have just see this question. I never experienced this last month and data are still flowing for me.
Hope this help

Reply:
Big coincidence, but just today it suddenly started reporting data again 😄 It was probably something specifically with my devices/cloud connection, because the Miele Android app has not had any contact with my Miele devices either. I'll make a new attempt and try to follow the reported energy consumption now again and see if it's more stable.

Reply:
The double consumption problem has happened twice since the API started working for me again, even with the revert in place.

<img width="1163" alt="bild" src="https://user-images.githubusercontent.com/1417619/146811008-d1151600-008d-4a4f-bf52-b303fc5b99b5.png">

My suspicion is that this may happen the first time the Miele device is started _after_ Home Assistant itself has been restarted. As long as Home Assistant is not restarted, the cached value in the Miele integration is returned until the Miele device is run the next time. But this check should handle that 😒 For some reason it doesn't seem to manage to do that any more.

```python
if not _is_running(device_status_value):
    self._cached_consumption = -1
    return 0
```

Looking at the raw event data in the Home Assistant recorder DB is just depressing. The first row is right before the dishwasher was started. The second row is the first scrape after it was powered on, and the third row is the third scrape. Row number 2 and 3 only differ in two places; the energy consumption and the water consumption 😞 So I honestly have no idea how to handle this in our integration..

```
MariaDB [home_assistant]> SELECT state, created, attributes FROM states WHERE entity_id = 'sensor.dishwasher_status' AND created > '2021-12-19 14:00:00' AND created < '2021-12-19 19:00:00';

...

| Off      | 2021-12-19 14:49:02.507613 | {"ProgramID":"ECO","rawProgramID":3,"programType":"Program","rawProgramType":0,"programPhase":"","rawProgramPhase":0,"dryingStep":"","rawDryingStep":null,"spinningSpeed":null,"rawSpinningSpeed":null,"ventilationStep":"","rawVentilationStep":null,"progress":0.0,"finishTime":"20:45","kickoffTime":"16:49","friendly_name":"Dishwasher Status"}                                                                                                                                                                                             
| In use   | 2021-12-19 14:49:12.492552 | {"ProgramID":"ECO","rawProgramID":3,"programType":"Program","rawProgramType":0,"programPhase":"Main wash","rawProgramPhase":1795,"dryingStep":"","rawDryingStep":null,"spinningSpeed":null,"rawSpinningSpeed":null,"ventilationStep":"","rawVentilationStep":null,"currentWaterConsumption":8.0,"currentWaterConsumptionUnit":"l","currentEnergyConsumption":0.7,"currentEnergyConsumptionUnit":"kWh","waterForecast":0.3,"energyForecast":0.3,"progress":0.0,"finishTime":"20:45","kickoffTime":"16:49","friendly_name":"Dishwasher Status"}    |
| In use   | 2021-12-19 14:49:17.516398 | {"ProgramID":"ECO","rawProgramID":3,"programType":"Program","rawProgramType":0,"programPhase":"Main wash","rawProgramPhase":1795,"dryingStep":"","rawDryingStep":null,"spinningSpeed":null,"rawSpinningSpeed":null,"ventilationStep":"","rawVentilationStep":null,"currentWaterConsumption":0.0,"currentWaterConsumptionUnit":"l","currentEnergyConsumption":0.0,"currentEnergyConsumptionUnit":"kWh","waterForecast":0.3,"energyForecast":0.3,"progress":0.0,"finishTime":"20:45","kickoffTime":"16:49","friendly_name":"Dishwasher Status"}    |
```

Reply:
Anyway, I still think that restoring the order of the checks as they were in my PR is the right thing to do. It at least reduces the risk for this happening. I can open a PR for it.

Reply:
If i understand correctly you got a double consumption because the statement was put up higher. Is that correct? I haven't seen this happen myself though, so just checking.

Reply:
Some of those higher consumptions have been caused by that yes, because my Miele devices seem to lose their connection to the cloud from time to time. But the ones I mentioned here today are not fixed by #91 unfortunately. I'm afraid those need to be fixed by Miele in the API.